### ISSUE TITLE
LGTM.com - false positive - Module is imported with both 'import' and 'import from'

### ISSUE THREAD

**Issue by LefterisJP** [github/codeql#6088]
**Description of the false positive**

At least from my eyes I have not been able to see a place where both `import` and `import from` are used for logging.

For built-in logging `import` is used, and for the projects own logging `import from`

**URL to the alert on the project page on LGTM.com**
https://lgtm.com/projects/g/rotki/rotki/snapshot/14ef94331652f0e3911d5f9c364abdeb93db4094/files/rotkehlchen/exchanges/independentreserve.py?sort=name&dir=ASC&mode=heatmap#x8492b90100d77b20:1

**Comment by tausbn** [comment]
Thank you for your report!

I think what's happening is the following:

1. Because `rotkehlchen/__main__.py` contains an `if __name__ == "__main__"` statement in the global scope, we assume that you may be running this file directly (i.e. `python __main__.py`).
2. When you run Python files directly, imports are resolved differently. In particular, if you run `__main__.py` directly, `import logging` will import the _local_ `logging.py` file in the same directory, and not the one in the standard library.
3. Because of this possibility, we conclude that the bare name `logging` is a possible name for `logging.py`. Under some circumstances you _will_ get that `import logging` resolves to that particular file.
4. However, we also note that this is a file that lives in a package, and so `rotkehlchen.logging` will _also_ resolve to that file, which ultimately results in the alerts.

There's still something odd in `import logging` resolving to `rotkehlchen/logging.py` from inside the `exchanges` directory. That should not happen. This will require further investigation.

However, even if that were fixed, you would still have an alert in `rotkehlchen/logging.py` saying that this file is importing itself (which it _is_, if you're running `__main__.py` directly). One solution to this would be to remove the `if __name__ == "__main__"` part, and replace it with just a call to `main()`. (After all, the whole purpose of the `__main__.py` file is to enable its contents to be executed using `python -m rotkehlchen`, so there's probably no need to double-guard it with the `if` statement.)

That being said, if writing `if __name__ == "__main__"` is common in `__main__.py` files, then we'll likely change the analysis to account for this fact. This should stop `logging.py` from getting the name `logging` and make all of the alerts go away.

**Comment by LefterisJP** [comment]
Hey thanks for the detailed answer.

> However, even if that were fixed, you would still have an alert in rotkehlchen/logging.py saying that this file is importing itself (which it is, if you're running __main__.py directly).

 Something similar is happening here:

https://lgtm.com/projects/g/rotki/rotki/snapshot/b7b0cb68263121c5f67b619b8c56c5d30d9fe4e8/files/rotkehlchen/logging.py#x803630daa00bc514:1

> That being said, if writing if __name__ == "__main__" is common in __main__.py files, then we'll likely change the analysis to account for this fact. This should stop logging.py from getting the name logging and make all of the alerts go away.

Yes this is indeed a very common practise. The reasons are explained very well here: https://stackoverflow.com/questions/419163/what-does-if-name-main-do

**Comment by tausbn** [comment]
> > However, even if that were fixed, you would still have an alert in rotkehlchen/logging.py saying that this file is importing itself (which it is, if you're running **main**.py directly).
> 
> Something similar is happening here:
> https://lgtm.com/projects/g/rotki/rotki/snapshot/b7b0cb68263121c5f67b619b8c56c5d30d9fe4e8/files/rotkehlchen/logging.py#x803630daa00bc514:1

Indeed, those alerts are ultimately caused by the same issue, the fact that `logging` may import itself. This is not strictly an error, by the way: if you do `python rotkehlchen/__main__.py`, then when `logging.py` is imported, it will import _itself_, and not the standard library `logging`, and this is likely not what you want.
 
> > That being said, if writing if **name** == "**main**" is common in **main**.py files, then we'll likely change the analysis to account for this fact. This should stop logging.py from getting the name logging and make all of the alerts go away.
> 
> Yes this is indeed a very common practise. The reasons are explained very well here: https://stackoverflow.com/questions/419163/what-does-if-name-main-do

I don't see any mention of the `__main__.py` file in the above link, so I'm not sure how the link is relevant to this discussion. 

Looking at just the `python/cpython` repo (which I would assume is authoritative on this subject), there are 33 `__main__.py` files. Of these, 23 just call the `main` function (or whatever) directly, and 10 of them have the `if __name__ ...` guard. Here's a link to an LGTM.com demonstrating this: https://lgtm.com/query/251953028916691429/

However, to my dismay, it seems that people _do_ write things like `import foo.__main__` (it even happens in the `cpython` repo!), so putting `if __name__ ...` in a `__main__.py` file can still make sense (even though to me it seems like a code smell to integrate `__main__.py` into the imports like this).

With that in mind, I have created a PR that fixes the problem here: https://github.com/github/codeql/pull/6091

Once it has been merged, the alerts should disappear from LGTM.com within a few weeks. :slightly_smiling_face:

**Comment by LefterisJP** [comment]
Thank you

**Comment by vedantmgoyal2009** [comment]
Closes #4879
